File: app/src/main/java/com/mithril/mobilegoldenleaf/Database/dao/ProductDao.java
package com.mithril.mobilegoldenleaf.Database.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.mithril.mobilegoldenleaf.models.Product;

import java.util.List;

@Dao
public interface ProductDao {

    @Insert
    long save(Product product);

    @Update
    void edit(Product product);

    @Query("SELECT * FROM Product")
    List<Product> all();


}

File: app/src/main/java/com/mithril/mobilegoldenleaf/database/dao/ProductDao.kt
package com.mithril.mobilegoldenleaf.database.dao

import androidx.room.*

import com.mithril.mobilegoldenleaf.models.Product

@Dao
interface ProductDao {

    @Insert
    fun save(product: Product): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun update(product: Product)

    @Query("SELECT * FROM Product")
    fun all(): List<Product>


    @Query("SELECT p.* FROM Product p JOIN Category c ON p.categoryId = c.id WHERE p.categoryId = :categoryId")
    fun productsWith(categoryId: Int): List<Product>


}
